ce/api/util.py

```diff
@@ -25,18 +25,21 @@ def get_units_from_netcdf_file(nc, variable):
     return nc.variables[variable].units
 
 
-def get_units_from_file_object(file_, varname):
+def get_units_from_file_object(file_, varname, ensemble_name):
```

Comment:
I think this is the wrong place to add the extra filtering functionality.

This function is passed a datafile and asked to return the units of the requested variable in that file. 

So for example if the user was seeking information about `tx90pETCCDI` with the `r1i1p1` run in the `ce_files` data collection ("ensemble") used by the Climate Explore Website, `get_units_from_run_object()` would collect all files that have the variable `tx90pETCCDI` and the run `r1i1p1`. Then it would get the set of all units from each file, in order to make sure they all had the same units and were comparable, and if so, the rest of the API would go about returning that comparison information. 

So in this case `get_units_from_file_object` is called on every file as long as that file has the following two properties:
* it contains some `tx90pETCCDI` data
* it has the run `r1i1p1`

So if you call this function with, say, `/storage/data/climate/downscale/CMIP5/BCCAQ/climdex/CanESM2_historical+rcp85_r1i1p1/tx90pETCCDI_yr_BCCAQ+ANUSPLIN300+CanESM2_historical+rcp85_r1i1p1_1950-2100.nc` - this file does have `tx90pETCCDI` data, and it does have  the `r1i1p1` run, but it is not in the `ce_files` group (because it's a weird 150 year file, not a 30-year summary file, and we don't display the 150 year files on the web because it takes one million years, but we do do some other things with them. This file also happens to have different units - the whole reason we want to filter by data collection in the first place is so that the existance of this weird 150 year file with goofy units doesn't affect unrelated requests for information from the `ce_files` group).

So when `get_units_from_file_object()` is what units are in this file, it iterates over the variables in the file (the first for loop, line 29) and over the data collections this file is a member of (the second for loop, line 30) but it never returns anything, because it is not a member of the `ce_files` data collection, so the second half of the line 31 `if` is never true. Instead it reaches the end of the function and raises an error, halting calculations, which is not what we want.

I think the easiest way forward would be to  move filtering-by-data-group up into `get_units_from_run_object()` or maybe `get_files_from_run_variable()`, though there are definitely other options, if you wanted to do something else.

Reply:
I'd also add that these kinds of operations on database objects can be extremely slow, as they issue separate queries to the database to access relation-based values like `file_.data_file_variables` and `dfv.ensembles`. These are convenient for programming but can be very inefficient. The better way is to issue queries with appropriate joins in the first place and rather walk through the database after a more limited query. The balance of course is between programming convenience and flexibility on the one hand and efficient use of the database on the other.